src/aeso_ast_infer_types.erl

```diff
@@ -862,17 +888,71 @@ infer1(Env, [{pragma, _, _} | Rest], Acc, Options) ->
     %% Pragmas are checked in check_modifiers
     infer1(Env, Rest, Acc, Options).
 
+check_implemented_interfaces([], _, _, _, _) ->
+    ok;
+check_implemented_interfaces([{contract_interface, _, IName, Extensions, Decls} | Interfaces], ConId, Impls, Acc, AllInterfaces) ->
+    case lists:member(name(IName), Acc) of
+        true ->
+            check_implemented_interfaces(Interfaces, ConId, Impls, Acc, AllInterfaces);
+        false ->
+            Unmatched = match_impls(Decls, ConId, name(IName), Impls),
+            NewInterfaces = [proplists:get_value(name(I), AllInterfaces) || I <- Extensions],
+            check_implemented_interfaces(Interfaces ++ NewInterfaces, ConId, Unmatched, [name(IName) | Acc], AllInterfaces)
+    end.
+
+match_impls([], _, _, Impls) ->
+    Impls;
+match_impls([{fun_decl, _, {id, _, FunName}, {fun_t, _, _, ArgsTypes, RetDecl}} | Decls], ConId, IName, Impls) ->
+    Match = fun({letfun, _, {id, _, FName}, Args, RetFun, _}) when FName == FunName ->
+                    length(ArgsTypes) == length(Args) andalso
+                    compare_types(RetDecl, RetFun) andalso
+                    lists:all(fun({T1, {typed, _, _, T2}}) -> compare_types(T1, T2) end,
+                              lists:zip(ArgsTypes, Args));
+               (_) -> false
+            end,
+    UnmatchedImpls = case lists:search(Match, Impls) of
+                         {value, V} ->
+                             lists:delete(V, Impls);
+                         false ->
+                             type_error({unimplemented_interface_function, ConId, IName, FunName}),
+                             Impls
+                     end,
+    match_impls(Decls, ConId, IName, UnmatchedImpls).
+
+
+-spec compare_types(T, T) -> boolean() when T :: utype() | [utype()].
+compare_types(Types1 = [_ | _], Types2 = [_ | _]) ->
```

Comment:
It's been a long time since I wrote this. I didn't know about `is_list/1` back then and thought this is the only way to check if a variable is a list.

Fixed here: 55bc58286e224ab54fff62a3b8ae2cd0ca87adcb